### PR TITLE
chore: break build if a class is not covered 

### PR DESCRIPTION
**Pull request by monperrus**
We should break the build if a class is not covered.

According to https://sonarqube.ow2.org/component_measures?id=fr.inria.gforge.spoon%3Aspoon-core&metric=coverage&view=list all classes are covered.

Yet, here Jacoco says that 8 classes are not covered: https://travis-ci.com/github/monperrus/spoon/jobs/396558655

    check (check-coverage) on project spoon-core: Coverage checks have not been met. See log for details

which ones?

### CODE REVIEW

**Comment by monperrus** [comment]
To get the list of not-covered classes.

First, generate the XML report of jacoco

    java -jar jacococli.jar report target/jacoco.exec --xml ./report.xml --sourcefiles src/main/java/ --classfiles target/classes/

Next, list the classes with 0 class coverage

    xmlstarlet fo -D  report.xml | xmlstarlet sel -t -v "//class/counter[@covered=0 and @type='CLASS']/../@name"

This gives:

```
spoon/support/gui/SpoonModelTree
spoon/support/gui/SpoonObjectFieldsTable
spoon/support/gui/SpoonTreeBuilder
spoon/support/gui/SpoonObjectFieldsTable$SpoonObjectTableModel
spoon/support/gui/SpoonTreeBuilder$1
spoon/support/gui/SpoonModelTree$1
spoon/support/gui/SpoonModelTree$2
spoon/support/gui/SpoonModelTree$3
spoon/support/gui/SpoonModelTree$4
spoon/support/gui/SpoonModelTree$5
spoon/support/gui/SpoonModelTree$6
spoon/support/gui/SpoonModelTree$7
spoon/support/reflect/declaration/CtElementImpl$1
spoon/support/util/EmptyClearableList$EmptyListIterator
spoon/support/compiler/jdt/JDTTreeBuilder$SpoonReferenceBinding
spoon/support/template/Parameters$1
```

Most of them are GUI classes meant to be excluded. And there are 4 corner cases of inner/anonymous classes, I'm looking at them.

**Comment by monperrus** [comment]
#3642 is for `spoon/support/reflect/declaration/CtElementImpl$1`

**Comment by monperrus** [comment]
#3643 is for `spoon/support/util/EmptyClearableList$EmptyListIterator`

**Comment by monperrus** [comment]
#3645 is for `spoon/support/compiler/jdt/JDTTreeBuilder$SpoonReferenceBinding`

**Comment by monperrus** [comment]
and #3646 for `spoon/support/template/Parameters$1`

**Comment by monperrus** [comment]
```
[INFO] --- jacoco-maven-plugin:0.8.6:check (check-coverage) @ spoon-core ---
[INFO] Loading execution data file /home/martin/martin-no-backup/spoon/spoon-core/target/jacoco.exec
[INFO] Analyzed bundle 'spoon-core' with 873 classes
[INFO] All coverage checks have been met.
```

File: pom.xml [hunk 1]
```diff
@@ -190,8 +184,33 @@
...
+                    </excludes>
+                    <rules>
+                    <rule implementation="org.jacoco.maven.RuleConfiguration">
+                      <element>BUNDLE</element>
+                      <limits>
+                        <limit implementation="org.jacoco.report.check.Limit">
+                          <counter>CLASS</counter>
+                          <value>MISSEDCOUNT</value>
+                          <!-- in theory we would like to have MISSEDCOUNT=0 but in some cases, Jacoco counts one phantom missed instruction, so ze go for one -->
+                          <maximum>7</maximum>
```
Comment (MartinWitt) [inline pom.xml:205]: shouldn't there be a comment for which instructions are phantom missed? This could help in the future, if the build fails.

**Comment by monperrus** [comment]
Making progress, now we have the faulty classes in the CI log. We needed to have `<element>CLASS</element>` (and not <s>`<element>BUNDLE</element>`</s>):

```
[WARNING] Rule violated for class spoon.support.visitor.replace.ReplacementVisitor.CtCompilationUnitDeclaredTypeReferencesReplaceListener: lines covered ratio is 0.0, but expected minimum is 0.1
[WARNING] Rule violated for class spoon.support.visitor.replace.ReplacementVisitor.CtCompilationUnitPackageDeclarationReplaceListener: lines covered ratio is 0.0, but expected minimum is 0.1
[WARNING] Rule violated for class spoon.support.visitor.replace.ReplacementVisitor.CtCompilationUnitDeclaredModuleReferenceReplaceListener: lines covered ratio is 0.0, but expected minimum is 0.1
[WARNING] Rule violated for class spoon.support.visitor.replace.ReplacementVisitor.CtCompilationUnitImportsReplaceListener: lines covered ratio is 0.0, but expected minimum is 0.1
[WARNING] Rule violated for class spoon.refactoring.MethodCallState: lines covered ratio is 0.0, but expected minimum is 0.1
[WARNING] Rule violated for class spoon.refactoring.MethodInvocationSearch: lines covered ratio is 0.0, but expected minimum is 0.1
[WARNING] Rule violated for class spoon.refactoring.CtDeprecatedRefactoring: lines covered ratio is 0.0, but expected minimum is 0.1
```

Thanks a lot to @marchof for the great help

**Comment by monperrus** [comment]
And now I understand why I had more coverage locally than on Travis, the refactoring tests are not executed in Travis because they test the Java version first in `RefactoringTest` and don't want to run with Java 8.

The solution is to move the verify to another job with Java 10 or 11.

**Comment by monperrus** [comment]
I'm done here

**Comment by nharrand** [comment]
Really neat!

**Pull request merged**